Add tooltips and run links to graph (#57)

rcowsill opened the pull request:
(Work in Progress for issue #57)

This PR adds tooltips and links to the relevant run for all filled areas of the graph:

![image](https://github.com/jenkinsci/build-time-blame-plugin/assets/42620235/6eea605d-8067-45a0-bfde-d20a1d7b9ac7)

I found a couple of problems while working on this:
1. Coordinates in imagemaps aren't scaled to match the dimensions of the image they're used on. If the image is scaled down then the imagemap areas don't line up with the graph.
2. The default StackedAreaRenderer has a bug which omits half of each area from the imagemap.

The "trend graph" in Jenkins core avoids those problems by:
1. Making the graph fixed-size, which guarantees that the imagemap will always line up correctly
2. Using a custom renderer (StackedAreaRenderer2) which doesn't have the imagemap generation bug

The simplest approach (used by this PR) is to follow the example of Jenkins core, making the graph fixed-size and reusing their renderer. It does mean that the graph no longer scales down at smaller window sizes, and the graph appearance changes a little.

I tried out an alternative approach if those downsides are unacceptable. That would keep the existing graph appearance and image scaling, but requires some external .js ([davidjbradshaw/image-map-resizer](https://github.com/davidjbradshaw/image-map-resizer)).

### Testing done

* All tests pass (including new ones for the custom renderer methods)
* Manually tested graph tooltip and link functionality using `mvn hpi:run`

```[tasklist]
### Remaining work
- [x] Update example image in the documentation to show tooltips
- [x] Add any additional tests (if required)
```

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

LTegtmeier commented:
> Making the graph fixed-size, which guarantees that the imagemap will always line up correctly

I'm happy with this compromise. I doubt that the dynamic size is worth any extra complexity.

LTegtmeier commented:
> Add any additional tests (if required)

If there's an easy way to test that the `BlameStackedAreaRenderer` generates a useful tooltip, it might be worthwhile but, if not I'm okay without new tests for this change  since it's basically just visual.

rcowsill commented:
Sounds good!

One thing I did notice after submitting this draft: the graph doesn't render properly when there's only one build result. StackedAreaRenderer2 draws nothing before the first build or after the last. If there's only one build it therefore renders nothing at all. The regular Jenkins trend graph has the same issue. I'm going to see if there's an easy workaround for it.

rcowsill committed:
Add tooltips and run links to graph (#57)

* Switch to a custom renderer with tooltip/URL generators and a
workaround for an image map generation bug in JFreeChart.

* Remove max-width CSS rule from trend graph

<map> coordinates don't scale to match images they're used by, so this
change makes the graph fixed-size.

* Set upper/lower domain axis margins to 0 and set graph width to 900px

This removes ~100px of empty space surrounding the plot area, which
reduces the impact of switching to a fixed-size graph.

* Add tests of new renderer methods

* Update build time trend documentation image

rcowsill commented:
I couldn't find a workaround to show results for just one build using StackedAreaRenderer2, so took another approach.

There's a simple workaround for the image map generation bug in JFreeChart, so I've used that instead. The bug only affects multi-polygon hotspots, so splitting up the hotspots to contain one polygon each avoids it.

Currently the workaround is in a separate class which BlameStackedAreaGenerator extends, but I'm happy to merge them if preferred.

I've also added tests for the new renderer methods, and updated the example graph image in the documentation.

LTegtmeier commented on src/test/groovy/org/jenkins/ci/plugins/buildtimeblame/analysis/BlameReportTest.groovy:
```diff
@@ -119,6 +119,68 @@ class BlameReportTest extends Specification {
+        toolTip != null
+        toolTip.contains("${buildNumber}")
+        toolTip.contains('Middle')
+        toolTip.contains('123')
```
It took me a while to figure out why we expect `123`. This might make that easier.
```suggestion
        toolTip.contains('123s')
```

LTegtmeier commented on src/main/groovy/org/jenkins/ci/plugins/buildtimeblame/analysis/MapFixStackedAreaRenderer.groovy:
```diff
@@ -0,0 +1,46 @@
+            int type = pi.currentSegment(coords)
+
+            switch (type) {
+                case PathIterator.SEG_MOVETO:
+                    subPath.moveTo(coords[0], coords[1])
+                    break
+                case PathIterator.SEG_LINETO:
+                    subPath.lineTo(coords[0], coords[1])
+                    break
```
How hard would it be to test these two cases? 

To be honest, I don't follow too much of how this works but, the results look good.

rcowsill replied:
These cases are indirectly covered by the 'should split compound paths' test. The `pathsMatch` argument constraint checks that moveTo/lineTo are getting called correctly while building the subPath eventually passed to `entityCollection.add`.

I could add another test with GeneralPath mocked, and interactions that directly confirm these cases are handled correctly.

LTegtmeier replied:
Oh, I see. Checking the results like your existing test is better than mocking. It just takes some IDE navigating to follow it.

LTegtmeier reviewed (commented):
Looking good! Thanks for adding this feature and including the tests.

I'm curious, how did you get the same data into the new example image?

rcowsill commented:
> I'm curious, how did you get the same data into the new example image?

I wrote an additional test to do that. It makes a BlameReport with that data, saves a graph .png of it and makes a HTML file using the image and map. After that it's a case of loading the HTML in a browser and screenshotting manually.

I can make a separate PR with that if you think it'd be useful.

LTegtmeier committed:
Clarify the elapsed seconds tooltip assertion

LTegtmeier commented:
> I wrote an additional test to do that. It makes a BlameReport with that data, saves a graph .png of it and makes a HTML file using the image and map. After that it's a case of loading the HTML in a browser and screenshotting manually.
> 
> I can make a separate PR with that if you think it'd be useful.

👍 It'd be nice to have an easy way to keep the documentation up to date.

The ideal thing is probably to set it up so that there's a job with realistic build data in the Jenkins server you can run with `hpi:run` to test the plugin. But, I'm not sure what that looks like and a quick test/script is almost as useful.
